### DISCUSSION TITLE
Lua -> Rust Function that accepts multiple UserData types as a parameter

### DISCUSSION THREAD

**Discussion by thomascft** [mlua-rs/mlua#336] [Q&A] [answered]
I'm working on a GTK4 widget system similar to something like [eww](https://github.com/elkowar/eww).
Right now I'm trying to create a function that will accept a widget UserData as an arg and set that widget as a windows child.
The widget can be any descendant of gtk::Widget which is what's causing my confusion. What would be the appropriate signature for this closure?
```
|lua, widget: ???| {


                ApplicationWindow::builder()
                    .application(app)
                    .child(&widget)
                    .build()
                    .present();

                Ok(())
            }
```

**Comment by khvzak** [comment]
You can use [`AnyUserData`](https://docs.rs/mlua/latest/mlua/struct.AnyUserData.html) type and downcast it to a desired/required type using [`AnyUserData::borrow`](https://docs.rs/mlua/latest/mlua/struct.AnyUserData.html#method.borrow)

If you don't know exact (final) type, then you can store types in Enum or try something like `Box<dyn IsA<Widget>>` (in this case you need to store types in Lua as trait objects too)

**Reply by thomascft** [reply]
Thanks for the quick response! `IsA<gtk::Widget>` isn't object safe, so I've created a trait object called `IsWidget` with a method that returns a `gtk::Widget`. I'm getting a UserDataTypeMismatch error while trying to cast the reference to a struct that wraps the widget. It's highly likely I just don't fully understand traits well enough. Here's what I've got so far. 
```rust
        let l_create_label = scope.create_function(|_, label: String|{
            let widget = gtk::Label::builder()
                .label(label)
                .build();
            Ok(Label(widget))
        }).unwrap();      

        let l_init = scope.create_function(|_:_, ud: AnyUserData|{
            let widget = ud.borrow::<Ref<'_, dyn IsWidget>>().unwrap();

            ApplicationWindow::builder()
                .application(app)
                .child(&widget.get_widget())
                .build()
                .present();
            Ok(())
        }).unwrap();
```

**Reply by thomascft** [reply]
I've made some progress. I'm not sure if the use of `dyn` really matters, but it compiles. 
```rust
        let l_create_label = scope.create_function(|_, label: String|{
            let label = Label(
                gtk::Label::builder()
                    .label(label)
                    .build()
            );

            lua.create_userdata::<Box<dyn IsWidget>>(Box::new(label))

        })?;      

         let l_create_button = scope.create_function(|_, label: String|{
            let button = Button(
                gtk::Button::builder()
                    .label(label)
                    .build()
            );

            lua.create_userdata::<Box<dyn IsWidget>>(Box::new(button))

        })?;
        let l_init = scope.create_function(|_, ud: mlua::AnyUserData|{
            
            let widget = ud.borrow::<Box<dyn IsWidget>>()?;

            let widget = widget.get_widget();
           // ...
```
```rust
pub trait IsWidget {
    fn get_widget(&self) -> gtk::Widget;
}
impl UserData for Box<dyn IsWidget> {}
```

**Reply by khvzak** [reply, marked as answer]
You can also optionally register `gtk::Widget` type directly with Lua using [`Lua::register_userdata_type`](https://docs.rs/mlua/latest/mlua/struct.Lua.html#method.register_userdata_type) and then create objects using [`Lua::create_any_userdata`](https://docs.rs/mlua/latest/mlua/struct.Lua.html#method.create_any_userdata)
